CoreData: Change fetchRequest to makeFetchRequest

davidrothera opened the pull request:
Fixes #725

Technically I can see this as a breaking change so added as such in the `CHANGELOG`

SwiftGen-Eve commented:
<!--
  0 Errors
  0 Warnings
  0 Messages
  5 Markdowns
-->

Hey 👋 I'm Eve, the friendly bot watching over SwiftGen 🤖

Thanks a lot for your contribution!



---

Seems like everything is in order 👍 You did a good job here! 🤝

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

djbe commented on CHANGELOG.md:
```diff
@@ -6,7 +6,8 @@
 
-_None_
+* CoreData: Changes `fetchRequest` auto-generated method to `createFetchRequest` to avoid ambiguous function usage.  
+  [David Rothera](https://github.com/davidrothera)
```
Please add a link to the pull request using the same format as other changelog entries.

davidrothera commented:
> You'll need to apply the same change to the swift 4 template.
> 
> I'm not 100% certain what the naming conventions are, and am very bad at it myself, so unless there's a better suggestion, let's go with `createFetchRequest()`.

Yeah I'm not sure on whether this is the correct name for the method either, it was either this or `makeFetchRequest` when I was thinking about it.

djbe commented:
@davidrothera thx for the quick PR!

Maybe we can avoid this being a breaking change, if we keep the old method name, and add a new method `createFetchRequest()`. We could maybe even add a `@available` annotation marking the old one as deprecated:

```swift
@available(*, deprecated, renamed: "createFetchRequest", message: "To avoid collisions with the less concrete method in `NSManagedObject`, please use `createFetchRequest()` instead.")
@nonobjc internal class func fetchRequest() -> NSFetchRequest<SecondaryEntity> {
  return NSFetchRequest<SecondaryEntity>(entityName: entityName)
}

@nonobjc internal class func createFetchRequest() -> NSFetchRequest< SecondaryEntity > {
  return NSFetchRequest<SecondaryEntity>(entityName: entityName)
}
```

davidrothera commented:
> @davidrothera thx for the quick PR!
> 
> Maybe we can avoid this being a breaking change, if we keep the old method name, and add a new method `createFetchRequest()`. We could maybe even add a `@available` annotation marking the old one as deprecated:
> 
> ```swift
> @available(*, deprecated, renamed: "createFetchRequest", message: "To avoid collisions with the less concrete method in `NSManagedObject`, please use `createFetchRequest()` instead.")
> @nonobjc internal class func fetchRequest() -> NSFetchRequest<SecondaryEntity> {
>   return NSFetchRequest<SecondaryEntity>(entityName: entityName)
> }
> 
> @nonobjc internal class func createFetchRequest() -> NSFetchRequest< SecondaryEntity > {
>   return NSFetchRequest<SecondaryEntity>(entityName: entityName)
> }
> ```

Thanks for the suggestion, yeah I would always prefer not having to break backwards compatibility. I'll get this changed shortly 👍

davidrothera commented:
I've added the method back in along with a deprecation notice, I updated the `CHANGELOG` to move it to the "New Features" section but not sure whether it would be best there or under one of the other headings.

djbe commented on Documentation/templates/coredata/swift5.md:
```diff
@@ -46,7 +46,7 @@ internal class MainEntity: NSManagedObject {
     return NSEntityDescription.entity(forEntityName: entityName, in: managedObjectContext)
   }
 
-  @nonobjc internal class func fetchRequest() -> NSFetchRequest<MainEntity> {
```
Please also make these changes in the swift 4 docs.

djbe commented:
Changelog location is ok for now, we can always move it later if really needed.

davidrothera commented:
Hey @djbe, have you had chance to look over the latest changes you requested?

djbe commented:
Allright, seems like those last changes cover everything.

One unfortunate bit is that I read through the API design guidelines:
> Begin names of factory methods with “make”, e.g. x.makeIterator().

So it should become `makeFetchRequest()` 😅 Luckily, we can find/replace now that the initial effort is done.
Feel free to rebase your commits if you find that easier.

davidrothera commented:
Good find and nice to know, there are probably some places in my own code which I should clean up in that case as well 😝

davidrothera commented:
Ok should be good to go now @djbe
